### PR TITLE
fix gamepad support

### LINKED ISSUE #43: D-pad doesn't work if using a joypad
D-pad doesn't work if using a joypad, analog l-stick works as d-pad instead (quick menu --> controls --> User 1 Device Type = RetroPad).

Keyboard works fine

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/50021173-d-pad-doesn-t-work-if-using-a-joypad?utm_campaign=plugin&utm_content=tracker%2F48210062&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F48210062&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by tcamargo**
Based on libretro-mame2016.
Should fix #43.

### CODE REVIEW

**Comment by tcamargo** [comment]
Any feedback? @twinaphex

**Comment by hizzlekizzle** [comment]
As mentioned on Discord, we should probably change this to use RETRO_DEVICE_JOYPAD instead of the keyboard, since the way 2016 does it is... weird.

**Comment by tcamargo** [comment]
Thx! I tried to use RETRO_DEVICE_JOYPAD without success. Ideas?

**Comment by Tatsuya79** [comment]
MAME stand-alone default is doing just the same: mapping direction on the analog left stick. (you can hit "delete" twice on a mapping to bring it back to default.

I can understand wanting to make it easier for users but I wonder if it's going to mess up the default mapping in other games that use the analog stick (changing that to dpad or having dpad directions mapped to several conflicting commands).

**Comment by tcamargo** [comment]
UDLR is mapped to both dpad and left analog. The original core code also mapped directionals to dpad, but it was not working.

**Comment by bigianb** [comment]
Given that the current state makes this core unusable on a cab with a discrete joystick (and this hopefully fixes that) ... does it have any bad side effects that would prevent merging it?

**Comment by tcamargo** [comment]
No side effect that I am aware of.

**Comment by tcamargo** [comment]
@bigianb as a workaround, for current core, you can map you discrete joy to left analog inputs in RA.

**Comment by bigianb** [comment]
> @bigianb as a workaround, for current core, you can map you discrete joy to left analog inputs in RA.

Ah ... yes of course ... just tested that and it works! Thank you (somewhat embarrassing not to figure that out ;-) )

**Comment by tcamargo** [comment]
It is also possible to workaround by using a custom cfg. You can generate one via MAME UI (TAB) and remap Px Up/Down/Left/Right.
```
<?xml version="1.0"?>
<!-- This file is autogenerated; comments and unknown tags will be stripped -->
<mameconfig version="10">
    <system name="default">
        <input>
            <port type="P1_JOYSTICK_UP">
                <newseq type="standard">
                    JOYCODE_1_BUTTON11
                </newseq>
            </port>
            <port type="P1_JOYSTICK_DOWN">
                <newseq type="standard">
                    JOYCODE_1_BUTTON12
                </newseq>
            </port>
            <port type="P1_JOYSTICK_LEFT">
                <newseq type="standard">
                    JOYCODE_1_BUTTON13
                </newseq>
            </port>
            <port type="P1_JOYSTICK_RIGHT>
                <newseq type="standard">
                    JOYCODE_1_BUTTON14
                </newseq>
            </port>
        </input>
    </system>
</mameconfig>
```
I am thinking about changing to HATxUP/DOWN/LEFT/RIGHT (input_uwp does that). Anyway, HAT is also not mapped by default. A cfg file is still needed.

**Comment by tcamargo** [comment]
I finally got an understating how inputs work in mame.

* Removed the keyboard device hack. I think it was used to avoid clashes. You can only have one button per input_item_id;
* Remapped d-pad to HAT input ids;
* Change default UDLR mapping to consider HAT ids.
* Cleaned-up input_retro.cpp a bit

Therefore, d-pad and analog inputs should work out of the box without the need for an external mapping file.

File: src/osd/modules/input/input_retro.cpp [hunk 1]
```diff
@@ -970,83 +956,71 @@ class joystick_input_retro : public retroinput_module
...
 				static_cast<input_item_id>(ITEM_ID_RYAXIS),
 				generic_axis_get_state<std::int32_t>,
 				&joystate[i].a2[1]);
-            
-         devinfo->device()->add_item(
+
+			devinfo->device()->add_item(
 				"L2",
-				static_cast<input_item_id>(ITEM_ID_RZAXIS),
+				static_cast<input_item_id>(ITEM_ID_ZAXIS),
```
Comment (Tatsuya79) [inline src/osd/modules/input/input_retro.cpp]: why are you changing the analog triggers? the default is good here testing in Out Run: R2 accelerates, L2 breaks.

Comment (tcamargo) [reply src/osd/modules/input/input_retro.cpp]: My mistake. I thought RZAXIS meant right. Just checked other input driver and you are right.

File: src/osd/modules/input/input_retro.cpp [hunk 2]
```diff
@@ -915,22 +907,16 @@ class retro_joystick_device : public event_based_device<KeyPressEventArgs>
...
 	}
 
 	void reset() override
 	{
-		int i,j;
-
-		for(i = 0; i < 4; i++)
-			for(j=0;j<RETRO_MAX_BUTTONS;j++)
-				joystate[i].button[j] = 0;	
+		memset(&joystate, 0, sizeof(joystate));
```
Comment (Tatsuya79) [inline src/osd/modules/input/input_retro.cpp:914]: isn't it going to reset only user 0 gamepad?

Comment (tcamargo) [reply src/osd/modules/input/input_retro.cpp:914]: All joystate entries.
```
#include <stdio.h>

typedef struct joystate_t
{
   int button[6];
   int a1[2];
   int a2[2];
   int a3[2];
} Joystate;

Joystate joystate[4];

int  main(int argc, char const *argv[]) {
	printf("size of joystate: %lu\n", sizeof(joystate));
	printf("size of Joystate: %lu\n", sizeof(Joystate));
	return 0;
}
```
```
$ ./a
size of joystate: 192
size of Joystate: 48
```

**Comment by Tatsuya79** [comment]
We don't have a default dpad bind for UI directions (Input (general) > User Interface > UI Up/Down/Left/Right). This should be added like for player controls.

As it is now you're losing UI directions in MAME GUI if you had something custom set there.

...but if we do it like for players with "Up or Dpad Up or JOY1 UP" I wonder if moving with the keyboard in the menu isn't going to do a ghost input (moving twice) as in RA the kb arrows are doing kb and also dpad directions together.

edit: well it should be fine as it's not doing double tap in game, sent a PR
https://github.com/libretro/mame/pull/127

**Comment by tcamargo** [comment]
It wont double tap. Currently, it has the same binding as normal UP/DOWN/LEFT/RIGHT. So, double tap would be happening now.

**Comment by 1985a** [comment]
Finally, I want to say thank you so much. I have been using a DS controller and I was a pain to play with the joystick. I'm from the old school so, always prefer to play with DPAD.

**Pull request merged**